diesel/src/mysql/connection/mod.rs

```diff
@@ -211,6 +211,34 @@ impl Connection for MysqlConnection {
     }
 }
 
+impl ConnectionWithReturningId for MysqlConnection {
+    type ReturnedId = u64;
+
+    fn execute_returning_id<T>(&mut self, source: &T) -> QueryResult<Self::ReturnedId>
+    where
+        T: QueryFragment<Self::Backend> + QueryId,
+    {
+        #[allow(unsafe_code)] // call to unsafe function
+        update_transaction_manager_status(
+            prepared_query(
+                &source,
+                &mut self.statement_cache,
+                &mut self.raw_connection,
+                &mut *self.instrumentation,
+            )
+            .and_then(|stmt| {
+                // we have not called result yet, so calling `execute` is
+                // fine
+                let stmt_use = unsafe { stmt.execute() }?;
+                Ok(unsafe { stmt_use.insert_id() })
```

Comment:
There should be a `Safety: explain how the invariants of the called function are uphold` comment here.

Additionally as far as I can see the `insert_id()` function states that it is only safe to call this function on insert statements, while we can call it on arbitrary statements here. That sounds like the current implementation is unsound.

In the end we want to replace the `T: QueryFragement<Self::Backend>` bound above with essentially `InsertStatement` to restrict the allowed statements to only insert statements.

Reply:
Will do

Reply:
> There should be a `Safety: explain how the invariants of the called function are uphold `comment here.

That's covered by the comment on lines 230-231, right?  This was copied+pasted directly from `execute_returning_count()` above and matches it precisely, the only change being the call to `stmt_use.insert_id()` instead of `stmt_use.affected_rows()`